Parser raises an exception if interval is set to "0.1s" but works fine with "100ms"

<!--
GitHub is reserved for bug reports and feature requests. The best place
to ask a general question is at the Elastic Discourse forums at
https://discuss.elastic.co. If you are in fact posting a bug report or
a feature request, please include one and only one of the below blocks
in your new issue.
-->

<!--
If you are filing a bug report, please remove the below feature
request block and provide responses for all of the below items.
-->

**Elasticsearch version**: `2.3.2`

**JVM version**: 

```
openjdk version "1.8.0_45-internal"
OpenJDK Runtime Environment (build 1.8.0_45-internal-b14)
OpenJDK 64-Bit Server VM (build 25.45-b02, mixed mode)
```

**OS version**: `Ubuntu 12.04.5 LTS (GNU/Linux 3.2.0-80-virtual x86_64)`

**Description of the problem including expected versus actual behavior**: date_histogram aggregations fail if interval is set to `0.1s` but work fine if set to `100ms`. I would say the problem comes from the parser and it doesn't have anything to do with date_histograms.

**Steps to reproduce**:
 1. Following query fails
```
curl -XGET 'http://localhost:9200/logstash-2016.05.23/_search?pretty' -d '
{
   "size" : 0,
   "aggs" : {
      "0" : {
         "aggs" : {
            "0" : {
               "date_histogram" : {
                  "field" : "@timestamp",
                  "interval" : "0.1s"
               }
            }
         },
         "filter" : {
            "query" : {
               "filtered" : {
                  "query" : {
                     "query_string" : {
                        "query" : "*"
                     }
                  },
                  "filter" : {
                     "bool" : {
                        "must" : [
                           {
                              "range" : {
                                 "@timestamp" : {
                                    "to" : 1464034121806,
                                    "from" : 1464034119663
                                 }
                              }
                           }
                        ]
                     }
                  }
               }
            }
         }
      }
   }
}
'
{
  "error" : {
    "root_cause" : [ {
      "type" : "illegal_argument_exception",
      "reason" : "Zero or negative time interval not supported"
    } ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query",
    "grouped" : true,
    "failed_shards" : [ {
      "shard" : 0,
      "index" : "logstash-2016.05.23",
      "node" : "z3wh9XcOQjCIyVjT2H-8qg",
      "reason" : {
        "type" : "illegal_argument_exception",
        "reason" : "Zero or negative time interval not supported"
      }
    } ]
  },
  "status" : 400
}
````
but it works fine if I change the `interval` field to the equivalent value `100ms` 

````
curl -XGET 'http://eln02.useb.cartodb.net:9200/logstash-2016.05.23/_search?pretty' -d '
{
   "size" : 0,
   "aggs" : {
      "0" : {
         "aggs" : {
            "0" : {
               "date_histogram" : {
                  "field" : "@timestamp",
                  "interval" : "100ms"
               }
            }
         },
         "filter" : {
            "query" : {
               "filtered" : {
                  "query" : {
                     "query_string" : {
                        "query" : "*"
                     }
                  },
                  "filter" : {
                     "bool" : {
                        "must" : [
                           {
                              "range" : {
                                 "@timestamp" : {
                                    "to" : 1464034121806,
                                    "from" : 1464034119663
                                 }
                              }
                           }
                        ]
                     }
                  }
               }
            }
         }
      }
   }
}
'
{
  "took" : 304,
  "timed_out" : false,
  "_shards" : {
    "total" : 5,
    "successful" : 5,
    "failed" : 0
  },

(...)
````
**Provide logs (if relevant)**:

````
2016-05-24 11:25:06,144][DEBUG][action.search            ] [eln02] All shards failed for phase: [query]
RemoteTransportException[[eln01][10.0.2.81:9300][indices:data/read/search[phase/query]]]; nested: SearchParseException[failed to parse search source [
{
   "size" : 0,
   "aggs" : {
      "0" : {
         "aggs" : {
            "0" : {
               "date_histogram" : {
                  "field" : "@timestamp",
                  "interval" : "0.1s"
               }
            }
         },
         "filter" : {
            "query" : {
               "filtered" : {
                  "query" : {
                     "query_string" : {
                        "query" : "*"
                     }
                  },
                  "filter" : {
                     "bool" : {
                        "must" : [
                           {
                              "range" : {
                                 "@timestamp" : {
                                    "to" : 1464034121806,
                                    "from" : 1464034119663
                                 }
                              }
                           }
                        ]
                     }
                  }
               }
            }
         }
      }
   }
}
]]; nested: IllegalArgumentException[Zero or negative time interval not supported];
Caused by: SearchParseException[failed to parse search source [
{
   "size" : 0,
   "aggs" : {
      "0" : {
         "aggs" : {
            "0" : {
               "date_histogram" : {
                  "field" : "@timestamp",
                  "interval" : "0.1s"
               }
            }
         },
         "filter" : {
            "query" : {
               "filtered" : {
                  "query" : {
                     "query_string" : {
                        "query" : "*"
                     }
                  },
                  "filter" : {
                     "bool" : {
                        "must" : [
                           {
                              "range" : {
                                 "@timestamp" : {
                                    "to" : 1464034121806,
                                    "from" : 1464034119663
                                 }
                              }
                           }
                        ]
                     }
                  }
               }
            }
         }
      }
   }
}
]]; nested: IllegalArgumentException[Zero or negative time interval not supported];
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:855)
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:654)
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:620)
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:371)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryTransportHandler.messageReceived(SearchServiceTransportAction.java:368)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryTransportHandler.messageReceived(SearchServiceTransportAction.java:365)
	at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33)
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:75)
	at org.elasticsearch.transport.netty.MessageChannelHandler$RequestHandler.doRun(MessageChannelHandler.java:300)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: Zero or negative time interval not supported
	at org.elasticsearch.common.rounding.TimeZoneRounding$Builder.<init>(TimeZoneRounding.java:62)
	at org.elasticsearch.common.rounding.TimeZoneRounding.builder(TimeZoneRounding.java:40)
	at org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramParser.parse(DateHistogramParser.java:184)
	at org.elasticsearch.search.aggregations.AggregatorParsers.parseAggregators(AggregatorParsers.java:198)
	at org.elasticsearch.search.aggregations.AggregatorParsers.parseAggregators(AggregatorParsers.java:176)
	at org.elasticsearch.search.aggregations.AggregatorParsers.parseAggregators(AggregatorParsers.java:103)
	at org.elasticsearch.search.aggregations.AggregationParseElement.parse(AggregationParseElement.java:60)
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:838)
````

--- a/core/src/main/java/org/elasticsearch/common/unit/TimeValue.java
+++ b/core/src/main/java/org/elasticsearch/common/unit/TimeValue.java
@@ -241,7 +241,7 @@ public class TimeValue implements Streamable {
             if (lowerSValue.endsWith("ms")) {
                 millis = (long) (Double.parseDouble(lowerSValue.substring(0, lowerSValue.length() - 2)));
             } else if (lowerSValue.endsWith("s")) {
-                millis = (long) Double.parseDouble(lowerSValue.substring(0, lowerSValue.length() - 1)) * 1000;
+                millis = (long) (Double.parseDouble(lowerSValue.substring(0, lowerSValue.length() - 1)) * 1000);
             } else if (lowerSValue.endsWith("m")) {
                 millis = (long) (Double.parseDouble(lowerSValue.substring(0, lowerSValue.length() - 1)) * 60 * 1000);
             } else if (lowerSValue.endsWith("h")) {

--- a/core/src/test/java/org/elasticsearch/common/unit/TimeValueTests.java
+++ b/core/src/test/java/org/elasticsearch/common/unit/TimeValueTests.java
@@ -83,6 +83,9 @@ public class TimeValueTests extends ESTestCase {
         assertEquals(new TimeValue(10, TimeUnit.SECONDS),
                      TimeValue.parseTimeValue("10S", null, "test"));
 
+        assertEquals(new TimeValue(100, TimeUnit.MILLISECONDS),
+                    TimeValue.parseTimeValue("0.1s", null, "test"));
+
         assertEquals(new TimeValue(10, TimeUnit.MINUTES),
                      TimeValue.parseTimeValue("10 m", null, "test"));
         assertEquals(new TimeValue(10, TimeUnit.MINUTES),

